pkg/actions/tools/nix/path.go

```diff
@@ -0,0 +1,29 @@
+package nix
+
+import (
+	"strings"
+
+	"github.com/rsteube/carapace"
+)
+
+// ActionPaths completes paths
+//
+// A path can be one of:
+//   - a local file path (default.nix)
+//   - an http/https URL (https://releases.nixos.org/../nixexprs.tar.xz
+//   - a channel: specifier (channel:nixos-22.05)
+//   - a local channel (<nixpkgs>)
+func ActionPaths() carapace.Action {
```

Comment:
This dispatches the completion based on the callback value but I don't know if it's more correct to just list everything. I'm mostly concerned about `ActionRemoteChannels` since that makes an http request.

Reply:
Yes, just thought so as well when I looked at it. The request is fine here, payload is not too heavy and can be done as multiparts completion (first complete prefixes and only make the http call on the next step).

Reply:
```diff
diff --git a/pkg/actions/tools/nix/path.go b/pkg/actions/tools/nix/path.go
index e3dbc8b2..1bacd46c 100644
--- a/pkg/actions/tools/nix/path.go
+++ b/pkg/actions/tools/nix/path.go
@@ -15,15 +15,17 @@ import (
 //   - a local channel (<nixpkgs>)
 func ActionPaths() carapace.Action {
 	return carapace.ActionCallback(func(c carapace.Context) carapace.Action {
-		arg := c.CallbackValue
-		if strings.HasPrefix(arg, "<") {
-			return ActionLocalChannels().Invoke(c).Prefix("<").Suffix(">").ToA()
-		} else if strings.HasPrefix(arg, "http:") || strings.HasPrefix(arg, "https:") {
-			return carapace.ActionValues()
-		} else if strings.HasPrefix(arg, "channel:") {
-			return ActionRemoteChannels().Invoke(c).Prefix("channel:").ToA()
+		batch := carapace.Batch(
+			ActionLocalChannels().Prefix("<").Suffix(">"),
+			carapace.ActionFiles(),
+		)
+
+		if !strings.HasPrefix(c.CallbackValue, "channel:") {
+			batch = append(batch, carapace.ActionValues("channel:").NoSpace())
 		} else {
-			return carapace.ActionFiles()
+			batch = append(batch, ActionRemoteChannels().Prefix("channel:"))
 		}
+
+		return batch.ToA()
 	})
 }
```